feat: add video conferencing option (Google Meet) to Google Calendar integration

akurungadam opened the pull request:
- Option to "Add Video Conferencing" in Event doctype
- Extend Google Calendar integration to add Google Meet link to calendar events 
    - insert and update APIs of Google Calendar now invoked with `conferenceDataVersion=1` [API Reference](https://developers.google.com/calendar/api/guides/create-events?hl=en#conferencing)
    - Add `conferenceData` as part of Google Calendar API
    - Add `attendees` based on Participants added to Event (adding attendees even if conferencing is not enabled, unlike the current implementation)
    - Event update (Add / Remove conferencing after syncing the event)
    - Show link to join video conference in Event document

Creating an Event with Video Conferencing enabled -
![Screen Shot 2022-08-11 at 2 03 34 PM](https://user-images.githubusercontent.com/3326959/185031032-d66f3fad-f0c9-4588-a071-9b6270acd873.png)

Google Meet link saved after inserting Event 
![Screen Shot 2022-08-11 at 5 32 59 PM](https://user-images.githubusercontent.com/3326959/185031223-24733f0a-7245-4a43-b414-cb0aa4285d72.png)

Show join link on dashboard
![2022-08-17 09 25 18](https://user-images.githubusercontent.com/3326959/185031797-f881f82b-2c42-434f-afa1-8d2eb8980769.jpg)

Docs, Google Calendar Integration [Draft]( https://frappeframework.com/docs/v14/user/en/guides/integration/google_calendar/edit-wiki?wiki_page_patch=7fa416deca )

akurungadam committed:
feat: add video conferencing (via google meet) for events syncing with google calendar

akurungadam committed:
fix: unset add_video_conferencing if sync_with_google_calendar not enabled

akurungadam committed:
fix: remove conference data if add_video_conferencing disabled during update

akurungadam committed:
fix: restrict add conferencing if event_category in Event, Meeting and Other

akurungadam committed:
fix: save meet link while pulling from calendar
updated response API Response comment with conference details
show google calendar section if pulled from google calendar

akurungadam committed:
feat: show link to join video conference with google meet

akurungadam committed:
fix(style): linter errors

akurungadam committed:
fix(style): linter errors

akurungadam committed:
fix: set conferenceDataVersion 1 only if google meet conferencing enabled
remove google meet link from event needs both conferenceData as null and conferenceDataVersion 1

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -462,9 +485,34 @@ def update_event_in_google_calendar(doc, method=None):
+		)
+
+		# if add_video_conferencing enabled or disabled during update, overwrite
+		frappe.db.set_value("Event", doc.name, {"google_meet_link": event.get("hangoutLink")})
```
can we put `update_modified=False` over here?

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -682,6 +730,44 @@ def get_recurrence_parameters(recurrence):
+				),
+				alert=True,
+				indicator="red",
+			)
```
this should ideally be taken out of the for loop ..as it will show a lot of messages. 

We could maintain a `dict` where we keep all the attendees whose emails we found and not found.

```py
{
"participant_email_found": set(),
"participant_email_not_found": set()
}
```

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -462,9 +485,34 @@ def update_event_in_google_calendar(doc, method=None):
+			conference_data_version = 1
+
+		if len(doc.event_participants):
+			event.update({"attendees": get_attendees(doc)})
```
if we remove *all* event participants, they still show up in the google calendar event

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -407,11 +409,31 @@ def insert_event_in_google_calendar(doc, method=None):
+	if len(doc.event_participants):
+		event.update({"attendees": get_attendees(doc)})
+
+	conference_data_version = 0
```
we don't really need this variable over here right? we can just say if doc.add_video_conferencing is there, we need a meet link.

akurungadam replied:
this i guess we should go with the variable as it wouldn't be so obvious if we use the add_video_conferencing flag to be passed as the version?

phot0n commented on frappe/desk/doctype/event/event.js:
```diff
@@ -41,6 +41,15 @@ frappe.ui.form.on("Event", {
 			__("Add Participants")
 		);
+
+		const [ends_on_date] = frm.doc.ends_on.split(" ");
+		if (frm.doc.google_meet_link && frappe.datetime.now_date() <= ends_on_date) {
```
this is only considering date?

if I had a meet from 4pm-6pm then I would get this header for the whole day (?)

#

Also, I'm getting errors if `ends_on` is empty.

<img width="719" alt="Screenshot 2022-09-15 at 8 28 54 PM" src="https://user-images.githubusercontent.com/32034600/190438003-791b7f20-b789-41c2-bf3e-87a8283ac243.png">

akurungadam replied:
> this is only considering date?
>
> if I had a meet from 4pm-6pm then I would get this header for the whole day (?)

yes, headline will be shown until end date. Google Calendar doesn't remove the "Join" option for past events too, and even after the meeting is completed, do you feel we'll keep it that way?

> Also, I'm getting errors if `ends_on` is empty.

this escaped, fixing this by using `starts_on` if `ends_on` is blank

phot0n replied:
> yes, headline will be shown until end date. Google Calendar doesn't remove the "Join" option for past events too, and even after the meeting is completed, do you feel we'll keep it that way?

seems okay now that I think about it - not a huge thing 👍

phot0n commented on frappe/integrations/doctype/google_calendar/google_calendar.py:
```diff
@@ -682,6 +730,44 @@ def get_recurrence_parameters(recurrence):
+
+		participant_doc = frappe.get_doc(participant.reference_doctype, participant.reference_docname)
+
+		if participant_doc.meta.has_field("user") and participant_doc.user:
+			attendees.append({"email": participant_doc.user})
+		elif participant_doc.meta.has_field("user_id") and participant_doc.user_id:
+			attendees.append({"email": participant_doc.user_id})
+		elif participant_doc.meta.has_field("email") and participant_doc.email:
+			attendees.append({"email": participant_doc.email})
```
this is crazy - what all doctypes are we currently supporting using this?

akurungadam replied:
yeah, this even i was doubtful.. it would be better we allow user to select email field in the participants table - will add this

phot0n replied:
discussed this with @ankush, maybe just do this when someone has linked contact document(s) ?

ankush replied:
Yes, Contact document implements linking contact info with arbitrary documents. 

Lets add another column and fetch primary email from the linked contract instead of user directly?  😄

akurungadam replied:
thanks! let me try and add that

akurungadam replied:
Hi @phot0n I've made the suggested changes. as you get the time, please take a look and let me know?

phot0n commented on frappe/desk/doctype/event/event.py:
```diff
@@ -55,6 +55,13 @@ def validate(self):
 		if self.sync_with_google_calendar and not self.google_calendar:
 			frappe.throw(_("Select Google Calendar to which event should be synced."))
 
+		if not self.sync_with_google_calendar or self.event_category not in [
+			"Event",
+			"Meeting",
+			"Other",
+		]:
+			self.add_video_conferencing = False
```
let `add_video_conferencing` show on all event categories? - we show the `sync_with_google_calendar` checkbox by default anyway (?)

```suggestion
		if not self.sync_with_google_calendar:
			self.add_video_conferencing = 0
```

phot0n commented on frappe/desk/doctype/event/event.json:
```diff
@@ -272,12 +275,27 @@
+  },
+  {
+   "default": "0",
+   "depends_on": "eval:(doc.sync_with_google_calendar && in_list([\"Event\", \"Meeting\", \"Other\"], doc.event_category))",
```
let `add_video_conferencing` show on all event categories? - we show the `sync_with_google_calendar` checkbox by default anyway (?)

```suggestion
   "depends_on": "eval:doc.sync_with_google_calendar",
```

phot0n reviewed (commented):
Hey, sorry for the delay. I've tested this out for a bit and broadly seems to work fine. have a few doubts though.

Thanks!!

akurungadam commented:
Thanks for taking this up, will come back on your review

akurungadam committed:
Update frappe/desk/doctype/event/event.py

show add_video_conferencing for events of all categories

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

akurungadam committed:
Update frappe/desk/doctype/event/event.json

show add_video_conferencing for events of all categories

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

akurungadam committed:
fix: handle empty ends_on, use starts_on

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

akurungadam committed:
fix: do not update Event modified time on db set meet link

Co-authored-by: Ritwik Puri <ritwikpuri5678@gmail.com>

akurungadam committed:
fix: add email field event participants table
set email from linked contact
google calendar event to use event participant email as attendees

akurungadam committed:
fix: attendees not cleared from google calendar event

akurungadam committed:
fix(style): linter suggestions

phot0n committed:
fix: make email field non-read only, called set_participants_email from before_save hook

phot0n commented:
Will merge upon successful CI run. Thanks for this & sorry for the delay 😅

phot0n committed:
test: update new no of fields in event doctype in test_fetch_to_customize
